state/azure/cosmosdb/cosmosdb.go

```diff
@@ -228,43 +226,214 @@ func (c *StateStore) Get(ctx context.Context, req *state.GetRequest) (*state.Get
 		return nil, err
 	}
 
-	item := CosmosItem{}
-	err = jsoniter.ConfigFastest.Unmarshal(readItem.Value, &item)
+	item, err := NewCosmosItemFromResponse(readItem.Value, c.logger)
 	if err != nil {
 		return nil, err
 	}
 
-	item.Etag = string(readItem.Response.ETag)
+	// We are sure this is a []byte if not nil
+	b, _ := item.Value.([]byte)
+	return &state.GetResponse{
+		Data: b,
+		ETag: ptr.Of(item.Etag),
+	}, nil
+}
 
-	if item.IsBinary {
-		if item.Value == nil {
-			return &state.GetResponse{
-				Data: make([]byte, 0),
-				ETag: ptr.Of(item.Etag),
-			}, nil
+// getMulti retrieves multiple items within a single partition, efficiently with a single query.
+func (c *StateStore) getMulti(ctx context.Context, pk string, req []*state.GetRequest) ([]state.BulkGetResponse, error) {
+	if len(req) == 0 {
+		return []state.BulkGetResponse{}, nil
+	}
+
+	// Build the query
+	var consistency azcosmos.ConsistencyLevel
+	keys := make([]string, len(req))
+	for i, r := range req {
+		keys[i] = r.Key
+
+		// Use the specified consistency if empty
+		// If there's a strong consistency, it overrides any eventual
+		if r.Options.Consistency == state.Strong {
+			consistency = azcosmos.ConsistencyLevelStrong
+		} else if r.Options.Consistency == state.Eventual && consistency == "" {
+			consistency = azcosmos.ConsistencyLevelEventual
 		}
+	}
 
-		bytes, decodeErr := base64.StdEncoding.DecodeString(item.Value.(string))
-		if decodeErr != nil {
-			c.logger.Warnf("CosmosDB state store Get request could not decode binary string: %v. Returning raw string instead.", decodeErr)
-			bytes = []byte(item.Value.(string))
+	// Execute the query
+	// Source for "ARRAY_CONTAINS": https://stackoverflow.com/a/44639998/192024
+	queryOpts := &azcosmos.QueryOptions{
+		QueryParameters: []azcosmos.QueryParameter{
+			{Name: "@keys", Value: keys},
+		},
+	}
+	if consistency != "" {
+		queryOpts.ConsistencyLevel = &consistency
+	}
+	pager := c.client.NewQueryItemsPager(
+		"SELECT * FROM r WHERE ARRAY_CONTAINS(@keys, r.id)",
+		azcosmos.NewPartitionKeyString(pk),
+		queryOpts,
+	)
+	result := make([]state.BulkGetResponse, len(req))
+	n := 0
+	for pager.More() {
+		queryResponse, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, err
 		}
 
-		return &state.GetResponse{
-			Data: bytes,
-			ETag: ptr.Of(item.Etag),
-		}, nil
+		for _, value := range queryResponse.Items {
+			item, err := NewCosmosItemFromResponse(value, c.logger)
+			if err != nil {
+				// If the error was while unserializing JSON, we don't have an ID, so exit right away
+				// This should never happen, hopefully
+				if item.ID == "" {
+					return nil, err
+				}
+				result[n] = state.BulkGetResponse{
+					Key:   item.ID,
+					Error: err.Error(),
+				}
+				n++
+				continue
+			}
+
+			// If the item contains a partition key, ensure it's the correct one
+			// This should never happen unless there's some issue
+			if item.PartitionKey != "" && item.PartitionKey != pk {
```

Comment:
Reviewers: this actually happened during testing: https://github.com/Azure/azure-sdk-for-go/issues/20543

Unclear why